source/common/network/connection_impl.cc

```diff
@@ -476,20 +476,14 @@ void ConnectionImpl::setBufferLimits(uint32_t limit) {
   // limits we err on the side of buffering more triggering watermark callbacks less often.
   //
   // Given the current implementation for straight up TCP proxying, the common case is reading
-  // |limit| bytes through the socket, passing |limit| bytes to the connection (triggering the high
-  // watermarks) and the immediately draining |limit| bytes to the socket (triggering the low
-  // watermarks). We avoid this by setting the high watermark to limit + 1 so a single read will
-  // not trigger watermarks if the socket is not blocked.
-  //
-  // If the connection class is changed to write to the buffer and flush to the socket in the same
-  // stack then instead of checking watermarks after the write and again after the flush it can
-  // check once after both operations complete. At that point it would be better to change the high
-  // watermark from |limit + 1| to |limit| as the common case (move |limit| bytes, flush |limit|
-  // bytes) would not trigger watermarks but a blocked socket (move |limit| bytes, flush 0 bytes)
-  // would result in respecting the exact buffer limit.
+  // |limit| bytes through the socket, passing |limit| bytes to the connection and the immediately
+  // draining |limit| bytes to the socket. Triggering the high watermarks and then immediately
+  // triggering the low watermarks would be expensive, but we narrowly avoid triggering high
+  // watermark when moving |limit| bytes through the connection because the high watermark
+  // computation checks if the size of the buffer exceeds the high watermark value.
   if (limit > 0) {
-    write_buffer_->setWatermarks(limit + 1);
-    read_buffer_->setWatermarks(limit + 1);
```

Comment:
I should mention that the +1 on the limit can result in a uint32_t overflow. One example where this happens is source/extensions/filters/network/dubbo_proxy/conn_manager.cc which passes in UINT32_MAX to setBufferLimits